### PR TITLE
[CI] add pattern components generator workflow

### LINKED ISSUE #40: [CI] Support for Generation of Pattern Components
**Current Behavior**
Meshery Adapter for NSM should have support for pattern components.

[Design spec reference](https://docs.google.com/document/d/1b8JAMzr3Rntu7CudRaYv6r6ccACJONAB5t7ISCaPNuA/edit#heading=h.lkm2k7nswsg)

**Desired Behavior**
Historical and version versions of NSM components should be created and embedded in this adapter repo.

**Implementation**

Create a GitHub workflow for getting the manifest from NGINX SM Helm Charts using the [kubeopenapi-jsonschema](https://github.com/layer5io/kubeopenapi-jsonschema) to read and generate json schemas from those manifests and then also generating workload definitions from those schemas.

A sample workflow is [here](https://github.com/DelusionalOptimist/meshery-istio/blob/gen-oam/.github/workflows/update-oam-defs.yml)

---
### Contributor Resources

The meshery.io website uses Jekyll and GitHub Pages. Site content is found under the [`master` branch](https://github.com/meshery/meshery.io/tree/master).
- See the [Contributing to Meshery.io Website](https://github.com/layer5io/meshery.io#contributing-to-the-mesheryio-website) section of the readme.md and other [contributing instructions](https://docs.meshery.io/project/contributing), too.
- See Meshery site designs in this [Figma designfile](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI?node-id=110%3A1). Fill-in the [Layer5 Community Member Form](https://layer5.io/newcomer) and join the [Layer5 Community Slack](http://slack.layer5.io) for access.

### PR DESCRIPTION
**Pull request by iamsdas**
**Description**

This PR fixes #40 

**Notes for Reviewers**

**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: .github/workflows/update-oam-defs.yml [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+          if [ -d templates/oam/workloads/workloads-${{ steps.glrt.outputs.release }} ]
+          then
+            echo "EXISTS=true" > $GITHUB_ENV
+          else
+            export "EXISTS=false" > $GITHUB_ENV
+          fi
+
+  generator:
+    name: Generate and push OAM definitions
+    needs: [check_nsm_version]
```
Comment (sayantan1413) [inline .github/workflows/update-oam-defs.yml:37]: Should all the nsm be replaced with nginx here? These two are seperate adaptor I guess.

Comment (iamsdas) [reply .github/workflows/update-oam-defs.yml:37]: I originally kept it as nginx-sm...then noticed the issue said "nsm adaptor"...then I replaced all the "nginx-sm" with "nsm" instead🤦

Comment (iamsdas) [reply .github/workflows/update-oam-defs.yml:37]: So before I do anything else, can someone confirm what name I should use?

Comment (sayantan1413) [reply .github/workflows/update-oam-defs.yml:37]: @DelusionalOptimist "nginx" would be fine right ?

Comment (warunicorn19) [reply .github/workflows/update-oam-defs.yml:37]: I think nginx would make more sense, since we already use `nsm` for the NSM adapter

Comment (iamsdas) [reply .github/workflows/update-oam-defs.yml:37]: `nginx` or `nginx-sm`?

Comment (DelusionalOptimist) [reply .github/workflows/update-oam-defs.yml:37]: Let's go with `nginx-sm` :+1:

**Comment by Shreyas220** [comment]
@iamsdas can you provide the link where the workloads are being generated

**Comment by iamsdas** [comment]
@Shreyas220 you can find them [here](https://github.com/meshery/meshery-nginx-sm/pull/50/commits/761f44afcc114c7de792cac40efde645010b1288)

File: .github/workflows/update-oam-defs.yml [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+          cd util
+          npm i; npm i nexe -g
+          make linux
+      - name: Generate Workload definitions
+        run: |
+          dirPath=../templates/oam/workloads/workloads-$NEW_NGINXSM_VERSION
+          cd util
+          mkdir -p $dirPath
+          meshName=nginxsm
+          smpMeshName=NGINXSM
```
Comment (DelusionalOptimist) [inline .github/workflows/update-oam-defs.yml:74]:
```suggestion
          smpMeshName=NGINX_SERVICE_MESH
```
SMP mesh name : )
https://github.com/service-mesh-performance/service-mesh-performance/blob/f5ffabafb5b4eed338123c4b10bb7942518f195a/spec/service_mesh.pb.go#L59

File: .github/workflows/update-oam-defs.yml [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+          template='{"apiVersion":"core.oam.dev/v1alpha1","kind":"WorkloadDefinition","metadata":{},"spec":{"definitionRef":{},"metadata":{"@type":"pattern.meshery.io/mesh/workload","meshVersion":"'$NEW__VERSION'","meshName":"'$smpMeshName'","k8sAPIVersion":null,"k8sKind":""}}}'
```
Comment (DelusionalOptimist) [inline .github/workflows/update-oam-defs.yml:75]: Missing meshVersion in generated workload definitions https://github.com/iamsdas/meshery-nginx-sm/commit/66757127548a2e7fc767ec3537831ee67a77e54d#diff-52c83c5775abd33ca3fbafc4abd55b6c5c2c1d41556d72983324f61805ba9940

**Review by DelusionalOptimist — CHANGES_REQUESTED** [review]

File: .github/workflows/update-oam-defs.yml [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            nameUpper=`echo $t  | tr -d '"'`
+            nameLower=`echo $t  | tr -d '"' | tr '[:upper:]' '[:lower:]'`
+            definitionRef=$(printf %s.nginxsm.meshery.layer5.io $nameLower)
+            apiVersion=$(./kubeopenapi-jsonschema --location ~/nginxsm.yaml -t yaml --filter '$[?(@.kind=="CustomResourceDefinition" && @.spec.names.kind=='$t')]..spec.versions[0]' --o-filter "$[].name" -o json | jq '.[]' | tr -d '"')
+            apiGroup=$(./kubeopenapi-jsonschema --location ~/nginxsm.yaml -t yaml --filter '$[?(@.kind=="CustomResourceDefinition" && @.spec.names.kind=='$t')]..spec' --o-filter "$[].group" -o json | jq '.[]' | tr -d '"')
+            ./kubeopenapi-jsonschema --location ~/nginxsm.yaml -t yaml --filter '$[?(@.kind=="CustomResourceDefinition" && @.spec.names.kind=='$t')]..openAPIV3Schema.properties.spec' --o-filter "$[]" -o json |\
+            jq '.[]' > $dirPath/$nameLower.nginxsm.meshery.layer5io.schema.json
+            echo $template |\
+            jq '
+            ."metadata"."name" = "'$(printf %s.nginxsm $nameUpper)'"
```
Comment (DelusionalOptimist) [inline .github/workflows/update-oam-defs.yml:87]:
```suggestion
            ."metadata"."name" = "'$(printf %s.nginx-sm $nameUpper)'"
```

File: .github/workflows/update-oam-defs.yml [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            | ."spec"."metadata"."k8sAPIVersion" = "'$(printf $apiGroup/$apiVersion $apiGroup $apiVersion)'"
+            | ."spec"."metadata"."k8sKind" = "'$nameUpper'"
+            | ."spec"."definitionRef"."name"="'$definitionRef'"' > $dirPath/$nameLower.nginxsm_definition.json
```
Comment (DelusionalOptimist) [inline .github/workflows/update-oam-defs.yml:90]:
```suggestion
            | ."spec"."definitionRef"."name"="'$definitionRef'"' > $dirPath/$nameLower.nginx-sm_definition.json
```

File: .github/workflows/update-oam-defs.yml [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+          crds=$(./kubeopenapi-jsonschema --location ~/nginxsm.yaml -t yaml --filter '$[?(@.kind=="CustomResourceDefinition")]' -o json --o-filter '$..["spec"]["names"]["kind"]' | jq '.[]')
+          for t in ${crds[@]}; do
+            nameUpper=`echo $t  | tr -d '"'`
+            nameLower=`echo $t  | tr -d '"' | tr '[:upper:]' '[:lower:]'`
+            definitionRef=$(printf %s.nginxsm.meshery.layer5.io $nameLower)
+            apiVersion=$(./kubeopenapi-jsonschema --location ~/nginxsm.yaml -t yaml --filter '$[?(@.kind=="CustomResourceDefinition" && @.spec.names.kind=='$t')]..spec.versions[0]' --o-filter "$[].name" -o json | jq '.[]' | tr -d '"')
+            apiGroup=$(./kubeopenapi-jsonschema --location ~/nginxsm.yaml -t yaml --filter '$[?(@.kind=="CustomResourceDefinition" && @.spec.names.kind=='$t')]..spec' --o-filter "$[].group" -o json | jq '.[]' | tr -d '"')
+            ./kubeopenapi-jsonschema --location ~/nginxsm.yaml -t yaml --filter '$[?(@.kind=="CustomResourceDefinition" && @.spec.names.kind=='$t')]..openAPIV3Schema.properties.spec' --o-filter "$[]" -o json |\
+            jq '.[]' > $dirPath/$nameLower.nginxsm.meshery.layer5io.schema.json
```
Comment (DelusionalOptimist) [inline .github/workflows/update-oam-defs.yml:84]:
```suggestion
            jq '.[]' > $dirPath/$nameLower.nginx-sm.meshery.layer5io.schema.json
```

File: .github/workflows/update-oam-defs.yml [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+          git clone https://github.com/layer5io/kubeopenapi-jsonschema util
+          cd util
+          npm i; npm i nexe -g
+          make linux
+      - name: Generate Workload definitions
+        run: |
+          dirPath=../templates/oam/workloads/workloads-$NEW_NGINXSM_VERSION
+          cd util
+          mkdir -p $dirPath
+          meshName=nginxsm
```
Comment (DelusionalOptimist) [inline .github/workflows/update-oam-defs.yml:73]:
```suggestion
          meshName=nginx-sm
```

File: .github/workflows/update-oam-defs.yml [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+          [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
+          nvm install v14.15.3
+          nvm install-latest-npm
+          git clone https://github.com/layer5io/kubeopenapi-jsonschema util
+          cd util
+          npm i; npm i nexe -g
+          make linux
+      - name: Generate Workload definitions
+        run: |
+          dirPath=../templates/oam/workloads/workloads-$NEW_NGINXSM_VERSION
```
Comment (DelusionalOptimist) [inline .github/workflows/update-oam-defs.yml:70]:
```suggestion
          dirPath=../templates/oam/workloads/$NEW_NGINXSM_VERSION
```

File: .github/workflows/update-oam-defs.yml [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+          cd util
+          mkdir -p $dirPath
+          meshName=nginxsm
+          smpMeshName=NGINX_SERVICE_MESH
+          template='{"apiVersion":"core.oam.dev/v1alpha1","kind":"WorkloadDefinition","metadata":{},"spec":{"definitionRef":{},"metadata":{"@type":"pattern.meshery.io/mesh/workload","meshVersion":"'$NEW_NGINXSM_VERSION'","meshName":"'$smpMeshName'","k8sAPIVersion":null,"k8sKind":""}}}'
+          crds=$(./kubeopenapi-jsonschema --location ~/nginxsm.yaml -t yaml --filter '$[?(@.kind=="CustomResourceDefinition")]' -o json --o-filter '$..["spec"]["names"]["kind"]' | jq '.[]')
+          for t in ${crds[@]}; do
+            nameUpper=`echo $t  | tr -d '"'`
+            nameLower=`echo $t  | tr -d '"' | tr '[:upper:]' '[:lower:]'`
+            definitionRef=$(printf %s.nginxsm.meshery.layer5.io $nameLower)
```
Comment (DelusionalOptimist) [inline .github/workflows/update-oam-defs.yml:80]:
```suggestion
            definitionRef=$(printf %s.nginx-sm.meshery.layer5.io $nameLower)
```

**Review by DelusionalOptimist — CHANGES_REQUESTED** [review]

**Review by DelusionalOptimist — APPROVED** [review]
Looks awesome :rocket:
Thanks @iamsdas : )

**Pull request merged**